File: editor/src/lib/evidenceTime.ts
export type ClockAnomalyReason =
  | "evidence-after-receipt"
  | "evidence-time-regression"
  | "append-receipt-regression"
  | "elapsed-time-divergence";

export type ClockAnomaly = {
  creditEventId: string;
  outcomeId: string;
  reason: ClockAnomalyReason;
  evidenceTime: string;
  appendReceiptAt: string;
  previousEvidenceTime: string | null;
  previousAppendReceiptAt: string | null;
};

export type EvidenceTime = {
  evidenceTime: number;
  appendReceiptAt: number;
};

export function parseEvidenceTime(evidenceTime: unknown, appendReceiptAt: unknown): EvidenceTime {
  const parsed = {
    evidenceTime: Date.parse(String(evidenceTime)),
    appendReceiptAt: Date.parse(String(appendReceiptAt)),
  };
  if (!Number.isFinite(parsed.evidenceTime) || !Number.isFinite(parsed.appendReceiptAt)) {
    throw new Error("evidence timestamps must be valid ISO timestamps");
  }
  return parsed;
}

export function evidenceAvailabilityTime(value: EvidenceTime): number {
  return Math.max(value.evidenceTime, value.appendReceiptAt);
}

export function clockAnomalies(
  current: EvidenceTime,
  input: {
    creditEventId: string;
    outcomeId: string;
    previous: EvidenceTime | null;
    delayed: boolean;
    maximumFutureSkewSeconds: number;
    maximumElapsedDivergenceSeconds: number;
  },
): ClockAnomaly[] {
  const reasons: ClockAnomalyReason[] = [];
  if (current.evidenceTime > current.appendReceiptAt + input.maximumFutureSkewSeconds * 1_000) {
    reasons.push("evidence-after-receipt");
  }
  if (input.previous !== null) {
    if (current.evidenceTime < input.previous.evidenceTime) reasons.push("evidence-time-regression");
    if (current.appendReceiptAt < input.previous.appendReceiptAt) reasons.push("append-receipt-regression");
    if (
      input.delayed
      && Math.abs(
        (current.evidenceTime - input.previous.evidenceTime)
        - (current.appendReceiptAt - input.previous.appendReceiptAt),
      ) > input.maximumElapsedDivergenceSeconds * 1_000
    ) {
      reasons.push("elapsed-time-divergence");
    }
  }
  return reasons.map((reason) => ({
    creditEventId: input.creditEventId,
    outcomeId: input.outcomeId,
    reason,
    evidenceTime: new Date(current.evidenceTime).toISOString(),
    appendReceiptAt: new Date(current.appendReceiptAt).toISOString(),
    previousEvidenceTime: input.previous === null
      ? null
      : new Date(input.previous.evidenceTime).toISOString(),
    previousAppendReceiptAt: input.previous === null
      ? null
      : new Date(input.previous.appendReceiptAt).toISOString(),
  }));
}

File: src/codaro/curriculum/evidenceTime.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import UTC, datetime, timedelta
from typing import Literal

from pydantic import BaseModel, ConfigDict


ClockAnomalyReason = Literal[
    "evidence-after-receipt",
    "evidence-time-regression",
    "append-receipt-regression",
    "elapsed-time-divergence",
]


class ClockAnomaly(BaseModel):
    model_config = ConfigDict(extra="forbid")

    creditEventId: str
    outcomeId: str
    reason: ClockAnomalyReason
    evidenceTime: str
    appendReceiptAt: str
    previousEvidenceTime: str | None = None
    previousAppendReceiptAt: str | None = None


@dataclass(frozen=True, slots=True)
class EvidenceTime:
    evidenceTime: datetime
    appendReceiptAt: datetime

    @classmethod
    def parse(cls, evidenceTime: str, appendReceiptAt: str) -> EvidenceTime:
        return cls(
            evidenceTime=_parseTimestamp(evidenceTime),
            appendReceiptAt=_parseTimestamp(appendReceiptAt),
        )

    @property
    def availabilityTime(self) -> datetime:
        return max(self.evidenceTime, self.appendReceiptAt)

    def anomalies(
        self,
        *,
        creditEventId: str,
        outcomeId: str,
        previous: EvidenceTime | None,
        delayed: bool,
        maximumFutureSkewSeconds: int,
        maximumElapsedDivergenceSeconds: int,
    ) -> list[ClockAnomaly]:
        reasons: list[ClockAnomalyReason] = []
        futureSkew = timedelta(seconds=maximumFutureSkewSeconds)
        if self.evidenceTime > self.appendReceiptAt + futureSkew:
            reasons.append("evidence-after-receipt")
        if previous is not None:
            if self.evidenceTime < previous.evidenceTime:
                reasons.append("evidence-time-regression")
            if self.appendReceiptAt < previous.appendReceiptAt:
                reasons.append("append-receipt-regression")
            if delayed:
                evidenceElapsed = self.evidenceTime - previous.evidenceTime
                receiptElapsed = self.appendReceiptAt - previous.appendReceiptAt
                divergence = abs(evidenceElapsed - receiptElapsed)
                if divergence > timedelta(seconds=maximumElapsedDivergenceSeconds):
                    reasons.append("elapsed-time-divergence")
        return [
            ClockAnomaly(
                creditEventId=creditEventId,
                outcomeId=outcomeId,
                reason=reason,
                evidenceTime=_formatTimestamp(self.evidenceTime),
                appendReceiptAt=_formatTimestamp(self.appendReceiptAt),
                previousEvidenceTime=(
                    _formatTimestamp(previous.evidenceTime)
                    if previous is not None
                    else None
                ),
                previousAppendReceiptAt=(
                    _formatTimestamp(previous.appendReceiptAt)
                    if previous is not None
                    else None
                ),
            )
            for reason in reasons
        ]


def _parseTimestamp(value: str) -> datetime:
    parsed = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if parsed.tzinfo is None:
        raise ValueError("evidence timestamp must include a timezone")
    return parsed.astimezone(UTC)


def _formatTimestamp(value: datetime) -> str:
    return value.astimezone(UTC).isoformat(timespec="milliseconds").replace("+00:00", "Z")
